### DISCUSSION TITLE
Keeping track of possible Transport API extensions.

### DISCUSSION THREAD

**Discussion by lovelydinosaur** [encode/httpx#1629] [General]
Let's try to collate any possible future Transport API extensions here.

Current state...

Request extensions:

* timeout - Dict[float]. With httpcore we currently support timeout configurations of "pool", "connect", "read", "write".

Response extensions:

* reason_phrase - bytes. Not used with HTTP/2+.
* http_version - bytes. Not always available, eg. with WSGI transports, mock transports. Client may default to b"HTTP/1.1".

Future areas that might be supported via Transport API extensions...

* Raw stream read/write interface in response to CONNECT/Upgrade. #1150
* Interface to support trailing headers. https://github.com/encode/httpx/issues/1149
* HTTP/2 server push support.
* Potentially stuff like socket address information / certificate info.
* Perhaps zero-copy/sendfile support?
* Low-level debug callbacks. https://github.com/encode/httpx/discussions/1529#discussioncomment-703162
* Support for HTTP/2 never-indexed headers. https://github.com/encode/httpx/issues/1627

**Comment by SamStephens** [comment]
For Trailing Headers, I think it's worth being clear there's two flavours here... receiving trailing headers while receiving a chunked response, as per https://github.com/encode/httpx/issues/1149, but also sending trailing headers as part of a chunked POST.

**Comment by karpetrosyan** [comment]
https://github.com/encode/httpcore/discussions/828 more context for HTTP/2 server push.
